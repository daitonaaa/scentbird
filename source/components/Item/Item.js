import PropTypes from 'prop-types';
import classNames from 'classnames';
import { connect } from 'react-redux';
import React, { Component } from 'react';

import './Item.scss';

import Child from 'components/Child';
import CreateForm from 'components/CreateForm';

import {
  setOpenId,
  resetOpenId,
  toggleItemChildsCheck,
  deleteItemAndSetCount,
} from 'actions/items';


export class Item extends Component {

  static propTypes = {
    id: PropTypes.number.isRequired,
    open: PropTypes.bool.isRequired,
    title: PropTypes.string.isRequired,
    childs: PropTypes.array.isRequired,

    setOpenId: PropTypes.func.isRequired,
    resetOpenId: PropTypes.func.isRequired,
    toggleItemChildsCheck: PropTypes.func.isRequired,
    deleteItemAndSetCount: PropTypes.func.isRequired,
  };

  componentWillReceiveProps(nextProps) {
    const oldChilds = this.props.childs;
    const {
      open,
      childs,
      resetOpenId,
    } = nextProps;

    if (!Array.isArray(childs) || !Array.isArray(oldChilds)) {

      return;
    }

    if (open && oldChilds.length && !childs.length) {
      resetOpenId();
    }
  }

  getAllCheckChildStatus() {
    const { childs } = this.props;

    let checkedAll = true;

    if (childs instanceof Array && childs.length) {
      childs.forEach(item => {
        if (!item.check) checkedAll = false;
      });

      return checkedAll;
    }

    return false;
  }

  handleChangeOpenId = () => {
    const {
      id,
      open,
      setOpenId,
      resetOpenId,
    } = this.props;

    open
      ? resetOpenId()
      : setOpenId(id);
  }

  handleDeleteItem = () => {
    const {
      id,
      open,
      resetOpenId,
      deleteItemAndSetCount,
    } = this.props;

    deleteItemAndSetCount(id);
    if (open) resetOpenId();
  }

  handleCheckAllChilds = () => {
    const {
      id,
      childs,
      toggleItemChildsCheck,
    } = this.props;
    const checkedStatus = !this.getAllCheckChildStatus();

    if (childs instanceof Array && childs.length) {
      toggleItemChildsCheck(id, checkedStatus);
    }
  }

  renderOpenChildBtn() {
    const {
      open,
      childs,
    } = this.props;

    if (childs instanceof Array && childs.length) return (
      <div
        onClick={this.handleChangeOpenId}
        className={classNames(
          'item__open-child-btn', { open }
        )}
      >
        {open ? 'Закрыть' : 'Открыть'}
      </div>
    );

    return (
      <div
        onClick={this.handleChangeOpenId}
        className={classNames(
          'item__open-child-btn empty', { open }
        )}
      >
        {open ? 'Закрыть' : 'Создать'}
      </div>
    );
  }

  renderChilds() {
    const {
      id,
      childs,
    } = this.props;

    if (childs instanceof Array && childs.length) {
      return childs.map(item =>
        <Child
          {...item}
          parentId={id}
          key={item.id}
        />
      );
    }
  }

  render() {
    const {
      id,
      open,
      title,
    } = this.props;
    const checkedAllChilds = this.getAllCheckChildStatus();

    return (
      <div className={classNames('item', { open })}>
        <div className="item__box">
          <div className="item__title">
            <span>{id}:</span> {title}
          </div>
          {this.renderOpenChildBtn()}
          <div
            onClick={this.handleCheckAllChilds}
            className={classNames(
              'item__checked-childs-btn', { checked: checkedAllChilds }
            )}
          >
            {checkedAllChilds ? 'Снять выделение' : 'Выделить все'}
          </div>
          <div
            className="item__delete"
            onClick={this.handleDeleteItem}
          >
            Удалить
          </div>
        </div>
        <div className={classNames('item__childs', { open })}>
          {this.renderChilds()}
          <CreateForm
            childPosition
            parentId={id}
          />
        </div>
      </div>
    );
  }
}


const mapDispatchToProps = {
  setOpenId,
  resetOpenId,
  toggleItemChildsCheck,
  deleteItemAndSetCount,
};

export default connect(
  ()=>({}),
  mapDispatchToProps
)(Item);
