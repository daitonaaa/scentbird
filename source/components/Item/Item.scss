.item {
  margin: 0 20px 30px;

  &-box {
    display: flex;
    justify-content: space-between;
    align-items: center;
    box-sizing: border-box;
    margin: 0 10px 10px;
    padding: 10px;
    background: #e9ffef;
    border: solid 1px #c8e8d1;

    &.open {
      background: #fdfaff;
      border: solid 1px #edd7fb;
    }
  }

  &__title {
    text-decoration: underline;
  }

  &__open-child-btn {
    padding: 5px 10px;
    width: 225px;
    background: #8aaf95;
    color: #fff;
    text-align: center;
    cursor: pointer;
    transition: .2s;

    &:hover, &.open {
      background: #a18aaf;
    }

    &.open:hover {
      background: #baa6c5;
    }

    &.empty {
      background: #cbead4;
    }
  }

  &__delete {
    padding: 5px 10px;
    background: #de9f9f;
    color: #fff;
    text-align: center;
    cursor: pointer;
    transition: .2s;

    &:hover {
      background: #ff6262;
    }
  }
}
