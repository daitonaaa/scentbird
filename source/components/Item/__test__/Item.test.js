import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Item } from 'components/Item/Item';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });


const defaultProps = {
  id: 1,
  open: false,
  title: 'something',

  setOpenId: ()=>{},
  resetOpenId: ()=>{},
  toggleItemChildsCheck: ()=>{},
  deleteItemAndSetCount: ()=>{},
};


describe('> > > COMPONENT - Item', () => {
  it('#create snapshot', () => {
    const output = shallow(
      <Item childs={[]} {...defaultProps} />
    );

    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('#test method: renderChilds => childs.map()', () => {
    const list = [
      { id: 1 }, { id: 2 },
    ];
    const output = shallow(
      <Item childs={list} {...defaultProps} />
    );

    expect(output.instance().renderChilds()).toBeDefined();
  });

  it('#test method: getAllCheckChildStatus => all check', () => {
    const list = [
      { id: 1, check: true },
      { id: 2, check: true },
    ];
    const output = shallow(
      <Item childs={list} {...defaultProps} />
    );

    expect(output.instance().getAllCheckChildStatus()).toBe(true);
  });

  it('#test method: getAllCheckChildStatus => one uncheck', () => {
    const list = [
      { id: 1, check: false },
      { id: 2, check: true },
    ];
    const output = shallow(
      <Item childs={list} {...defaultProps} />
    );

    expect(output.instance().getAllCheckChildStatus()).toBe(false);
  });

  it('#test method: renderOpenChildBtn', () => {
    const output = shallow(
      <Item childs={[]} {...defaultProps} />
    );

    expect(output.find('.item__open-child-btn').text()).toBe('Создать');
  });
});
