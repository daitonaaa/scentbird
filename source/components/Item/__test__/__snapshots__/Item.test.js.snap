// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`> > > COMPONENT - Item #create snapshot 1`] = `
<div
  className="item"
>
  <div
    className="item__box"
  >
    <div
      className="item__title"
    >
      <span>
        1
        :
      </span>
       
      something
    </div>
    <div
      className="item__open-child-btn empty"
      onClick={[Function]}
    >
      Создать
    </div>
    <div
      className="item__checked-childs-btn"
      onClick={[Function]}
    >
      Выделить все
    </div>
    <div
      className="item__delete"
      onClick={[Function]}
    >
      Удалить
    </div>
  </div>
  <div
    className="item__childs"
  >
    <Connect(CreateForm)
      childPosition={true}
      parentId={1}
    />
  </div>
</div>
`;
